### PR TITLE
Issue #38: add aiProcess_RemoveComponent post-processing

### PR DESCRIPTION
**Pull request by corot**
This PR fix the particular problem I describe in the issue #38, but given the wide use of this package, someone more familiar with it should verify it doesn't brake any other usescases.

### CODE REVIEW

File: test/test_loaded_meshes.cpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2008, Willow Garage, Inc.
```
Comment (davetcoleman) [inline test/test_loaded_meshes.cpp]: Seems like you should put your name here ;-)

File: test/test_loaded_meshes.cpp [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+{
+public:
+  CompareMeshVsPrimitive()
+  {
+  }
+
+  void SetUp()
+  {
+    // WARN1: we skip the sphere because it takes ages for each test, around 20 seconds each
+    // (possible bug?)
```
Comment (davetcoleman) [inline test/test_loaded_meshes.cpp]: IDK if this is a bug, but maybe create an issue noting this?

File: test/test_loaded_meshes.cpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    // WARN2: meshes created by createMeshFromShape for cylinders and cones use a
+    // magic number to set the smoothness (now is 100):
+    // https://github.com/corot/geometric_shapes/blob/patch-1/src/mesh_operations.cpp#L510)
+    // The STL meshes have been built with a more or less equivalent smoothness, but
+    // if this matching breaks, randomized tests can fail in some unlikely situations
```
Comment (davetcoleman) [inline test/test_loaded_meshes.cpp]: I don't think we want to add flakey tests - should we just use the cube as a test since it has no curves that might causing smoothing differences?

File: test/test_loaded_meshes.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    }
+    EXPECT_TRUE(found) << "No point inside the meshes was found (very unlikely)";
+  }
+}
+
+TEST_F(CompareMeshVsPrimitive, IntersectsRay)
+{
+  // Random rays must intersect both meshes nearly at the same points
+
+  SUCCEED() << "Test disabled by now, as it fails in around 1% cases";
```
Comment (davetcoleman) [inline test/test_loaded_meshes.cpp]: What would it take to enable this test? Should this whole function just be removed?

File: test/test_loaded_meshes.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    computeShapeBoundingSphere(loaded_ms, center2, radius2);
+
+    EXPECT_NEAR(radius1,     radius2,     0.001);
+    EXPECT_NEAR(center1.x(), center2.x(), 0.001);
+    EXPECT_NEAR(center1.y(), center2.y(), 0.001);
+    EXPECT_NEAR(center1.z(), center2.z(), 0.001);
+  }
+}
+
+TEST_F(CompareMeshVsPrimitive, BoxVertexCount)
```
Comment (davetcoleman) [inline test/test_loaded_meshes.cpp:193]: Great test! :-)

File: test/test_loaded_meshes.cpp (hunk 4 above)
Comment (corot) [inline test/test_loaded_meshes.cpp]: Maybe just choose some "nice" rays testing corner cases, instead of just random rays.
However, random rays should always work, if both meshes are exactly the same and with the same rotation. I wonder if one is slightly rotated relative to the other... this, together with conic shapes discretization can explain the discrepancies. I guess increasing the resolution used to create the conics will reduce the 1% of failures, but the value is hardcoded on shapes constructors

Comment (davetcoleman) [inline test/test_loaded_meshes.cpp]: > Maybe just choose some "nice" rays testing corner cases, instead of just random rays.

+1 to this, so that we can just move forward with merging this longstanding PR

File: test/test_loaded_meshes.cpp (hunk 2 above)
Comment (corot) [inline test/test_loaded_meshes.cpp]: Maybe I should... but would be great if someone with more experience with the library can corroborate my findings

File: test/test_loaded_meshes.cpp (hunk 3 above)
Comment (corot) [inline test/test_loaded_meshes.cpp]: Kind of agree. But I fell a bit frustrated with the wasted effort when I realized the resolution problems that i just wanted to solve them. But for sure, don't make any sense to add brittle tests! If you and @jacquelinekay want, I will just drop all but cube tests

File: test/test_loaded_meshes.cpp (hunk 2 above)
Comment (v4hn) [inline test/test_loaded_meshes.cpp]: Please create an issue with a simple test-case so that we can reproduce and analyze it.
There is no sense in adding warnings about possible library bugs into the code of the library...

File: test/test_loaded_meshes.cpp (hunk 4 above)
Comment (davetcoleman) [inline test/test_loaded_meshes.cpp]: ping @corot have you had any more experience with this change and PR? have you been using these changes since you first opened the PR with no problems?

Comment (davetcoleman) [reply test/test_loaded_meshes.cpp]: ping

Comment (corot) [reply test/test_loaded_meshes.cpp]: Sorry man, I was on holidays/very busy (depending on the week). Yes, I use my fork with primitives generated both with blender and with OpenSCAD and in both cases this PR saves the day on octomap clearing.
I'll review/apply your comments next week, as soon as I can get some spare time!

Comment (davetcoleman) [reply test/test_loaded_meshes.cpp]: @corot, you put a lot of work into this so i'd love to see this merged.

**Comment by corot** [comment]
Finally, I got some spare time. Well, I have plainly removed all your sources of complains :neckbeard:

If merged, I'll repeat for Kinetic.

I'll open an issue on the long computing time for spheres, if that still happen

**Comment by davetcoleman** [comment]
Your tests are failing:

```
[==========] Running 8 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 8 tests from CreateMesh
[ RUN      ] CreateMesh.stl
[==========] Running 5 tests from 1 test case.
[       OK ] CreateMesh.stl (1 ms)
[ RUN      ] CreateMesh.daeNoUp
[----------] Global test environment set-up.
/root/ws_moveit/src/geometric_shapes/test/test_create_mesh.cpp:55: Failure
Value of: mesh->vertices[0 + 1]
  Actual: 1
Expected: 0
[  FAILED  ] CreateMesh.daeNoUp (0 ms)
[ RUN      ] CreateMesh.daeYUp
[----------] 5 tests from CompareMeshVsPrimitive
[       OK ] CreateMesh.daeYUp (1 ms)
[ RUN      ] CreateMesh.daeZUp
[ RUN      ] CompareMeshVsPrimitive.ContainsPoint
/root/ws_moveit/src/geometric_shapes/test/test_create_mesh.cpp:55: Failure
Value of: mesh->vertices[0 + 1]
  Actual: 1
Expected: 0
[  FAILED  ] CreateMesh.daeZUp (0 ms)
[ RUN      ] CreateMesh.daeXUp
/root/ws_moveit/src/geometric_shapes/test/test_create_mesh.cpp:57: Failure
Value of: mesh->vertices[3 + 0]
  Actual: -1
Expected: 0
[  FAILED  ] CreateMesh.daeXUp (0 ms)
[ RUN      ] CreateMesh.daeNoUnit
/root/ws_moveit/src/geometric_shapes/test/test_create_mesh.cpp:55: Failure
Value of: mesh->vertices[0 + 1]
  Actual: 1
Expected: 0
[  FAILED  ] CreateMesh.daeNoUnit (1 ms)
[ RUN      ] CreateMesh.dae1M
/root/ws_moveit/src/geometric_shapes/test/test_create_mesh.cpp:55: Failure
Value of: mesh->vertices[0 + 1]
  Actual: 1
Expected: 0
[  FAILED  ] CreateMesh.dae1M (0 ms)
[ RUN      ] CreateMesh.dae10M
/root/ws_moveit/src/geometric_shapes/test/test_create_mesh.cpp:55: Failure
Value of: mesh->vertices[0 + 1]
  Actual: 1
Expected: 0
[  FAILED  ] CreateMesh.dae10M (0 ms)
[----------] 8 tests from CreateMesh (3 ms total)
[----------] Global test environment tear-down
[==========] 8 tests from 1 test case ran. (3 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] CreateMesh.daeNoUp
[  FAILED  ] CreateMesh.daeZUp
[  FAILED  ] CreateMesh.daeXUp
[  FAILED  ] CreateMesh.daeNoUnit
[  FAILED  ] CreateMesh.dae1M
[  FAILED  ] CreateMesh.dae10M
```

**Comment by corot** [comment]
Are you sure? Mines are CompareMeshVsPrimitive. :confused:
When I rerun just my tests I get all OK

```
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from CompareMeshVsPrimitive
[ RUN      ] CompareMeshVsPrimitive.ContainsPoint
[       OK ] CompareMeshVsPrimitive.ContainsPoint (1 ms)
[ RUN      ] CompareMeshVsPrimitive.IntersectsRay
[       OK ] CompareMeshVsPrimitive.IntersectsRay (0 ms)
[ RUN      ] CompareMeshVsPrimitive.BoundingSphere
[       OK ] CompareMeshVsPrimitive.BoundingSphere (0 ms)
[ RUN      ] CompareMeshVsPrimitive.BoxVertexCount
[       OK ] CompareMeshVsPrimitive.BoxVertexCount (1 ms)
[ RUN      ] CompareMeshVsPrimitive.BoxTriangleCount
[       OK ] CompareMeshVsPrimitive.BoxTriangleCount (0 ms)
[----------] 5 tests from CompareMeshVsPrimitive (2 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (2 ms total)
[  PASSED  ] 5 tests.
```

I think the ones failing are CreateMesh, that have been recently added by @de-vri-es

**Comment by davetcoleman** [comment]
Ah I see. I just retriggered the build to see if its a flaky test. Hopefully @de-vri-es can look at this

**Comment by de-vri-es** [comment]
Hmm, it might be that the added post processing step is altering the mesh, making the unit test fail. I'll have to investigate in more detail.

/edit: If this is the case that means the unit tests I wrote should be improved. Right now they assume the order of vertices isn't modified, for example. I suspect that may be causing the problem.

File: src/mesh_operations.cpp [hunk 6]
```diff
@@ -238,14 +238,31 @@ Mesh* createMeshFromBinary(const char *buffer, std::size_t size, const Eigen::Ve
...
   const aiScene* scene = importer.ReadFileFromMemory(reinterpret_cast<const void*>(buffer), size,
                                                      aiProcess_Triangulate            |
                                                      aiProcess_JoinIdenticalVertices  |
-                                                     aiProcess_SortByPType,
-                                                     hint.c_str());
-  if (!scene)
+                                                     aiProcess_SortByPType            |
+                                                     aiProcess_RemoveComponent        |
+                                                     aiProcess_OptimizeGraph          |
+                                                     aiProcess_OptimizeMeshes, assimp_hint.c_str());
```
Comment (de-vri-es) [inline src/mesh_operations.cpp]: `ai_OptimizeGraph` and `aiProcess_OptimizeMeshes` were moved to a separate post-processing step below in order to clear the root transform of the imported mesh (see #52 for more details).

This change adds them back to the first post processing step. Probably happened during a merge conflict?

**Comment by de-vri-es** [comment]
Ah, the test failure is actually legit. More details in the line comments. Yay for unit tests :dancers:

File: src/mesh_operations.cpp [hunk 7]
```diff
@@ -238,14 +238,31 @@ Mesh* createMeshFromBinary(const char *buffer, std::size_t size, const Eigen::Ve
...
+  if (scene)
+    return createMeshFromAsset(scene, scale, assimp_hint);
+  else
```
Comment (de-vri-es) [inline src/mesh_operations.cpp]: This also looks like a left-over from a merge conflict. The function shouldn't return yet unless `scene == NULL`.

**Comment by de-vri-es** [comment]
Note that another test will fail:

```
[ RUN      ] MeshPointContainment.Pr2Forearm
geometric_shapes/test/test_point_inclusion.cpp:263: Failure
Value of: 2338
Expected: ms->vertex_count
Which is: 504
[  FAILED  ] MeshPointContainment.Pr2Forearm (4 ms)
```

It looks like that assertion is too strict here, since this PR is supposed to cut down the number of vertices. Might be replaced with `ASSERT_TRUE(ms != NULL)` (which should probably be there anyway before dereferencing `ms`).

**Comment by v4hn** [comment]
On Thu, Sep 22, 2016 at 01:33:49PM -0700, Maarten de Vries wrote:

> It looks like that assertion is too strict here, since this PR is supposed to cut down the number of vertices.

I would prefer to simply update the number of vertices here.
Adapting tests is perfectly valid in pull-requests :)

Michael

**Comment by de-vri-es** [comment]
> I would prefer to simply update the number of vertices here.

But that would mean relying on the details of a post-processing step from an external library in the unit test. I'd argue that unit tests should only test one thing, which in this case is the point inclusion API. Correct loading of meshes would be something for a different unit test (which to some degree we do have now, although for a very simple mesh).

**Comment by corot** [comment]
My bad,  mea culpa. I fu... up the merge, sorry a lot.
Now is fixed, I hope: I suppose the PR now properly applies to the current code. MEsh creation test now PASSED. However, still a test fails:

Value of: 2338
Expected: ms->vertex_count
Which is: 504
[  FAILED  ] MeshPointContainment.Pr2Forearm (4 ms)
[----------] 1 test from MeshPointContainment (4 ms total)

[----------] 1 test from MergeBoundingSpheres
[ RUN      ] MergeBoundingSpheres.MergeTwoSpheres
[       OK ] MergeBoundingSpheres.MergeTwoSpheres (0 ms)
[----------] 1 test from MergeBoundingSpheres (0 ms total)

[----------] Global test environment tear-down
[==========] 16 tests from 7 test cases ran. (4 ms total)
[  PASSED  ] 15 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] MeshPointContainment.Pr2Forearm

 1 FAILED TEST

**Comment by v4hn** [comment]
On Fri, Sep 23, 2016 at 01:26:52AM -0700, Jorge Santos Simón wrote:

> Now is fixed, I hope: I suppose the PR now properly applies to the current code. MEsh creation test now PASSED.

Thanks!

> However, still a test fails:
> 
> Value of: 2338
> Expected: ms->vertex_count
> Which is: 504

That was to be expected :)
After looking at the test in detail I agree with @de-vri-es .
I filed #58 to fix this.

**Comment by davetcoleman** [comment]
I think you need to rebase to get https://github.com/ros-planning/geometric_shapes/pull/58 to fix Travis for this PR

**Comment by davetcoleman** [comment]
I just retriggered Travis to make sure this was still a good PR, but I think its ready @corot yes?

**Comment by davetcoleman** [comment]
All checks pass

**Comment by corot** [comment]
all fine for me

**Comment by v4hn** [comment]
for the record: you also cherry-picked to kinetic-devel

**Comment by davetcoleman** [comment]
via this pull request https://github.com/ros-planning/geometric_shapes/pull/60, yes
